File: src/Models/SistemaDeReservas.java
package Models;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class SistemaDeReservas {
    private List<Usuario> usuarios;
    private List<Sala> salas;

    public SistemaDeReservas() {
        this.usuarios = new ArrayList<>();
        this.salas = new ArrayList<>();
    }

    public List<Usuario> getUsuarios() {
        return usuarios;
    }

    public List<Sala> getSalas() {
        return salas;
    }

    public void carregarUsuarios(String caminhoArquivo) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(caminhoArquivo));
            String linha;
            while ((linha = br.readLine()) != null) {
                String[] partes = linha.split(";");
                if (partes.length == 4) {
                    String tipo = partes[0].trim();
                    String nome = partes[1].trim();
                    String id = partes[2].trim();
                    String email = partes[3].trim();

                    Usuario usuario;
                    if (tipo.equalsIgnoreCase("Aluno")) {
                        usuario = new Aluno(nome, email, id);
                        System.out.println("Aluno carregado: " + usuario.getNome());
                    } else if (tipo.equalsIgnoreCase("Professor")) {
                        usuario = new Professor(nome, email, id);
                        System.out.println("Professor carregado: " + usuario.getNome());
                    } else {
                        throw new IllegalArgumentException("Tipo de usuário desconhecido: " + tipo);
                    }

                    usuarios.add(usuario);

                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public void carregarSalas(String caminhoArquivo) {
        try {
            BufferedReader br = new BufferedReader(new FileReader(caminhoArquivo));
            String linha;
            while ((linha = br.readLine()) != null) {
                String idSala = linha.trim();
                if (!idSala.isEmpty()) {
                    Sala sala = new Sala(idSala);
                    salas.add(sala);
                    System.out.println("Sala carregada: " + sala.getIdSala());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public int getNumeroSalas() {
        return  salas.size();
    }

    public void carregarReservas(String caminhoArquivo) {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm");
        try {
            BufferedReader br = new BufferedReader(new FileReader(caminhoArquivo));
            String linha;
            while ((linha = br.readLine()) != null) {
                String[] partes = linha.split(";");
                if (partes.length == 5) {
                    String idReserva = partes[0].trim();
                    LocalDateTime dataHoraInicio = LocalDateTime.parse(partes[1].trim(), formatter);
                    LocalDateTime dataHoraFim = LocalDateTime.parse(partes[2].trim(), formatter);
                    String idSala = partes[3].trim();
                    String idUsuario = partes[4].trim();

                    Sala sala = salas.stream()
                            .filter(s -> s.getIdSala().equals(idSala))
                            .findFirst()
                            .orElseThrow(() -> new IllegalArgumentException("Sala não encontrada: " + idSala));

                    Reserva reserva = new Reserva(idReserva, idUsuario, idSala, dataHoraInicio, dataHoraFim);
                    sala.reservar(reserva);
                    System.out.println("Reserva carregada: " + reserva.getIdReserva());
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    public Usuario buscarUsuarioPorId(String id) {
        return usuarios.stream()
                .filter(usuario -> usuario.getId().equals(id))
                .findFirst()
                .orElse(null);
    }

    public List<Reserva> ListarReservasDoUsuario(String idUsuario) {
        List<Reserva> reservasDoUsuario = new ArrayList<>();
        Usuario usuario = buscarUsuarioPorId(idUsuario);

        if (usuario == null) {
            System.out.println("Usuário não encontrado.");
        }

        System.out.println("Reservas do usuário " + usuario.getNome() + ":");
        for (Sala sala : salas) {
            List<Reserva> reservas = sala.getReservas();
            for (Reserva reserva : reservas) {
                if (reserva.getIdUsuario().equals(idUsuario)) {
                    System.out.println(reserva);
                    reservasDoUsuario.add(reserva);
                }
            }
        }
        return reservasDoUsuario;
    }

    //buscar reseervas  por periodo usando lambda
    public List<Reserva> bucarReservasPorPeriodo(LocalDateTime dataInicio, LocalDateTime dataFim) {
        List<Reserva> reservasNoPeriodo = new ArrayList<>();
        for (Sala sala : salas) {
            List<Reserva> reservas = sala.getReservas();
            for (Reserva reserva : reservas) {
                if ((reserva.getDataHoraInicio().isAfter(dataInicio) || reserva.getDataHoraInicio().isEqual(dataInicio)) &&
                        (reserva.getDataHoraFim().isBefore(dataFim) || reserva.getDataHoraFim().isEqual(dataFim))) {
                    reservasNoPeriodo.add(reserva);
                }
            }
        }
        return reservasNoPeriodo;
    }

    public void cancelarReserva(String idReserva, Usuario solicitante) {
        for (Sala sala : salas) {
            List<Reserva> reservas = sala.getReservas();
            for (Reserva reserva : reservas) {
                if (reserva.getIdReserva().equals(idReserva)) {
                    if (reserva.getIdUsuario().equals(solicitante.getId()) || solicitante instanceof Professor) {
                        sala.cancelar(idReserva);
                        System.out.println("Reserva " + idReserva + " cancelada com sucesso.");
                        return;
                    } else {
                        System.out.println("Usuário não tem permissão para cancelar esta reserva.");
                        return;
                    }
                }
            }
        }
    }
    //gerar relataorio
    public String gerarRelatorio() {
        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("dd/MM/yyyy HH:mm");
        StringBuilder relatorio = new StringBuilder();
        relatorio.append("=== Relatório Detalhado de Reservas ===\n");
        for (Sala sala : salas) {
            relatorio.append("Sala ID: ").append(sala.getIdSala()).append("\n");
            List<Reserva> reservas = sala.getReservas();
            if (reservas.isEmpty()) {
                relatorio.append("Nenhuma reserva feita.\n");
            } else {
                relatorio.append("Total de Reservas: ").append(reservas.size()).append("\n");
                for (Reserva reserva : reservas) {
                    Usuario usuario = buscarUsuarioPorId(reserva.getIdUsuario());
                    relatorio.append("Reserva ID: ").append(reserva.getIdReserva()).append("\n")
                            .append("Usuário: ").append(usuario != null ? usuario.getNome() : "Desconhecido").append("\n")
                            .append("Início: ").append(reserva.getDataHoraInicio().format(formatter)).append("\n")
                            .append("Fim: ").append(reserva.getDataHoraFim().format(formatter)).append("\n")
                            .append("-----------------------------\n");
                }
            }
        }
        return relatorio.toString();
    }
}

File: refactor-python/Models/SistemaDeReservas.py
from datetime import datetime

from Models.Aluno import Aluno
from Models.Professor import Professor
from Models.Reserva import Reserva
from Models.Sala import Sala


class SistemaDeReservas:
    def __init__(self):
        self.usuarios = []
        self.salas = []

    def carregarUsuarios(self, caminho):
            try:
                with open(caminho, 'r', encoding='utf-8') as file:
                    for linha in file:
                        linha = linha.strip()
                        if not linha:
                            continue
                        partes = linha.split(';')
                        if len(partes) == 4:
                            tipo = partes[0].strip()
                            nome = partes[1].strip()
                            id = partes[2].strip()
                            email = partes[3].strip()
                            if tipo.lower() == 'aluno':
                                usuario = Aluno(nome, id, email)
                            elif tipo.lower() == 'professor':
                                usuario = Professor(nome, id, email)
                            else:
                                print(f"Tipo de usuário desconhecido: {tipo}")
                                continue
                            self.usuarios.append(usuario)
            except IOError as e:
                print(f"Arquivo de usuários não encontrado.{e}")

    def carregarSalas(self, caminho):
        try:
            with open(caminho, 'r', encoding='utf-8') as file:
                for linha in file:
                    idSala = linha.strip()
                    if idSala != '':
                        sala = Sala(idSala)
                        self.salas.append(sala)
                        print (f"Sala {idSala} carregada com sucesso.")
        except IOError as e:
            print(f"Arquivo de salas não encontrado.{e}")


    def carregarReservas(self, caminho):
        formato_data = "%d/%m/%Y %H:%M"
        try:
            with open(caminho, 'r', encoding='utf-8') as arquivo:
                for linha in arquivo:
                    linha = linha.strip()
                    if not linha:
                        continue
                    partes = linha.split(';')
                    if len(partes) == 5:
                        idReserva = partes[0].strip()
                        dataInicio = partes[1].strip()
                        dataFim = partes[2].strip()
                        idSala = partes[3].strip()
                        idUsuario = partes[4].strip()

                        dataHoraInicio = datetime.strptime(dataInicio.strip(), "%d/%m/%Y %H:%M")
                        dataHoraFim = datetime.strptime(dataFim.strip(), "%d/%m/%Y %H:%M")
                        sala = next((s for s in self.salas if s.getIdSala() == idSala), None)
                        if not sala:
                            raise ValueError(f"Sala com ID {idSala} não encontrada")
                        reserva = Reserva(idReserva, idUsuario, idSala, dataHoraInicio, dataHoraFim)
                        sala.reservar(reserva)
                        print(f'Reserva {idReserva} carregada com sucesso.')
        except IOError as e:
            print(f"Arquivo de reservas não encontrado.{e}")

    def buscarUsuarioPorId(self, id):
        for usuario in self.usuarios:
            if usuario.getId() == id:
                return usuario
        return None

    def listarReservasDoUsuario(self, idUsuario):
        reservasDoUsuario = []
        usuario = self.buscarUsuarioPorId(idUsuario)
        if usuario == None:
            print("Usuário não encontrado")
        for sala in self.salas:
            for reserva in sala.getReservas():
                if reserva.getIdUsuario() == idUsuario:
                    reservasDoUsuario.append(reserva)
        return reservasDoUsuario

    def buscarReservasPorPeriodo(self,  dataInicio, dataFim):
        reservasNoPeriodo = []
        for sala in self.salas:
            for reserva in sala.getReservas():
                if reserva.getDataHoraInico() >= dataInicio and reserva.getDataHoraFim() <= dataFim:
                    reservasNoPeriodo.append(reserva)
        return reservasNoPeriodo

    def cancelarReserva(self, idReserva, solicitante):
        for sala in self.salas:
            for reserva in sala.reservas:
                if reserva.getIdUsuario() == solicitante.getId() or isinstance(solicitante, Professor):
                    if reserva.getIdReserva() == idReserva:
                        sala.cancelar(idReserva)
                        print("Reserva cancelaada com sucesso")
                        return True
                    else:
                        print("O usuario não tem permissão para cancelar esta reserva, ou a reserva não existe")
                        return False

    def gerarRelatorio(self, salas):
        formatter = "%d/%m/%Y %H:%M"
        relatorio = ["=== Relatório Detalhado de Reservas ===\n"]

        for sala in salas:
            relatorio.append(f"Sala ID: {sala.getIdSala()}")
            reservas = sala.getReservas()
            if not reservas:
                relatorio.append("Nenhuma reserva feita.\n")
            else:
                relatorio.append(f"Total de Reservas: {len(reservas)}")
                for reserva in reservas:
                    usuario = self.buscarUsuarioPorId(reserva.getIdUsuario())
                    nome_usuario = usuario.nome if usuario else "Desconhecido"
                    relatorio.append(
                        f"Reserva ID: {reserva.id_reserva}\n"
                        f"Usuário: {nome_usuario}\n"
                        f"Início: {reserva.inicio.strftime(formatter)}\n"
                        f"Fim: {reserva.fim.strftime(formatter)}\n"
                        "-----------------------------"
                    )

        return "\n".join(relatorio)

    def getNumeroSalas(self):
        return len(self.salas)

    def getUsuarios(self):
        return self.usuarios

    def getSalas(self):
        return self.salas